diff --git a/go.mod b/go.mod
index ced285f..95a1267 100644
--- a/go.mod
+++ b/go.mod
@@ -2,4 +2,6 @@ module k8s.io/klog/v2
 
 go 1.21
 
+replace github.com/go-logr/logr => ../../../staging/github.com/go-logr/logr
+
 require github.com/go-logr/logr v1.4.1
diff --git a/go.sum b/go.sum
index 0934eaa..e69de29 100644
--- a/go.sum
+++ b/go.sum
@@ -1,2 +0,0 @@
-github.com/go-logr/logr v1.4.1 h1:pKouT5E8xu9zeFC39JXRDukb6JFQPXM5p5I91188VAQ=
-github.com/go-logr/logr v1.4.1/go.mod h1:9T104GzyrTigFIr8wt5mBrctHMim0Nb2HLGrmQ40KvY=
diff --git a/klog.go b/klog.go
index 319ffbe..04a8c21 100644
--- a/klog.go
+++ b/klog.go
@@ -739,6 +739,16 @@ func (l *loggingT) printDepth(s severity.Severity, logger *logWriter, filter Log
 	l.printWithInfos(buf, file, line, s, logger, filter, depth+1, args...)
 }
 
+func (l *loggingT) printDepthWithoutHeader(s severity.Severity, logger *logWriter, filter LogFilter, depth int, args ...interface{}) {
+	if false {
+		_ = fmt.Sprint(args...) //  // cause vet to treat this function like fmt.Print
+	}
+
+	buf, file, line := l.header(s, depth)
+	buf.Reset() // Reset header.
+	l.printWithInfos(buf, file, line, s, logger, filter, depth+1, args...)
+}
+
 func (l *loggingT) printWithInfos(buf *buffer.Buffer, file string, line int, s severity.Severity, logger *logWriter, filter LogFilter, depth int, args ...interface{}) {
 	// If a logger is set and doesn't support writing a formatted buffer,
 	// we clear the generated header as we rely on the backing
@@ -853,6 +863,24 @@ func (l *loggingT) printS(err error, s severity.Severity, depth int, msg string,
 	buffer.PutBuffer(b)
 }
 
+// printSRaw prints a log message with no header and no formatting.
+func (l *loggingT) printSRaw(err error, s severity.Severity, depth int, msg string, keysAndValues ...interface{}) {
+	// Only create a new buffer if we don't have one cached.
+	b := buffer.GetBuffer()
+	// The message is always quoted, even if it contains line breaks.
+	// If developers want multi-line output, they should use a small, fixed
+	// message and put the multi-line output into a value.
+	b.WriteString(strconv.Quote(msg))
+	var errKV []interface{}
+	if err != nil {
+		errKV = []interface{}{"err", err}
+	}
+	serialize.FormatKVs(&b.Buffer, errKV, keysAndValues)
+	l.printDepthWithoutHeader(s, nil, nil, depth+1, &b.Buffer)
+	// Make the buffer available for reuse.
+	buffer.PutBuffer(b)
+}
+
 // SetOutput sets the output destination for all severities
 func SetOutput(w io.Writer) {
 	logging.mu.Lock()
@@ -1762,6 +1790,14 @@ func ExitfDepth(depth int, format string, args ...interface{}) {
 	logging.printfDepth(severity.FatalLog, logging.logger, logging.filter, depth, format, args...)
 }
 
+// PrintSDepth is similar to InfoSDepth,
+// but never print header and formatting, like enable --skip_headers.
+func (v Verbose) PrintSDepth(depth int, msg string, keysAndValues ...interface{}) {
+	if v.enabled {
+		logging.printSRaw(nil, severity.InfoLog, depth+1, msg, keysAndValues...)
+	}
+}
+
 // LogFilter is a collection of functions that can filter all logging calls,
 // e.g. for sanitization of arguments and prevent accidental leaking of secrets.
 type LogFilter interface {
diff --git a/klogr.go b/klogr.go
index 6204c7b..bb05ef0 100644
--- a/klogr.go
+++ b/klogr.go
@@ -17,6 +17,8 @@ limitations under the License.
 package klog
 
 import (
+	"bytes"
+	"bufio"
 	"github.com/go-logr/logr"
 
 	"k8s.io/klog/v2/internal/serialize"
@@ -101,5 +103,14 @@ func (l klogger) WithCallDepth(depth int) logr.LogSink {
 	return &l
 }
 
+func (l klogger) Write(p []byte) (int, error) {
+	v := VDepth(l.callDepth+1, Level(0))
+	s := bufio.NewScanner(bytes.NewReader(p))
+	for s.Scan() {
+		v.PrintSDepth(l.callDepth+1, s.Text(), l.values...)
+	}
+	return len(p), s.Err()
+}
+
 var _ logr.LogSink = &klogger{}
 var _ logr.CallDepthLogSink = &klogger{}
